### PR TITLE
[Feature] Table downloads async

### LINKED ISSUE #8956: 🐛 Download All Candidates from pool table gives an empty file the first time you click
## 🐛 Bug

If you select All on the Pool Candidates table and then click "Download CSV" you will get a file with a header but no data rows. 

If you click again (and on subsequent clicks) you will get the correct file.

## 🦋 Expected Behaviour

It should work the first time!

## 📋 Steps to Reproduce

1. Log in as admin
2. Go to the Pool Candidates table
3. Select All
4. Click Download CSV and check the file, it will only be a header line
5. Repeat step 4. It should contain the data now.

## 📸 Screenshot

Add a screenshot (if possible).

## 🙋‍♀️ Proposed Solution

CSV file is getting generated before the GetSelectedPoolCandidates returns with data.
Any time clicking on download after that query is returned, file is generated properly.
We just need to wait for that query to finish before generating the file.

Only fetch data after the Download CSV button is clicked. Only create the file after the data is fetched. Show a spinner while this is happening.

## ✅ Acceptance Criteria

A set of assumptions which, when tested, verify that the bug was addressed.

- [ ] Clicking on Download CSV after selecting all candidates should download a file with all rows of data the first time.
- [ ] Same goes for printing
- [ ] Show a spinner while data is being fetched

### PR DESCRIPTION
**Pull request by esizer**
🤖 Resolves #8956 

## 👋 Introduction

This fixes an issue where downloads were empty on the first attempt. It achieved this by making the data query asynchronous. 

### Problem/Solution

There seemed to be a race condition where we were loading the data at the same time as generating either the CSV or print document. This refactors things to avoid this. We now load the data on click, awaiting it and then generating the document. This has some pros and cons, however I believe the pros outweigh the cons here.

#### Pros

- No more race conditions
- Data is only loaded when it is needed, not before

#### Cons

- Since we are no longer pre-loading data, it is a little longer wait when clicking an action

### CSV

While the existing `react-csv` we were using purported to support async handlers, it did not seem to work and there were many issues on their repo about this. Coupled with the fact that the library had not been updated in 2 years, I switched to the more consistently maintained `react-csv-downloader` instead.

Unfortunately, this did require some refactoring to how we download CSVs.

## 🧪 Testing

Assist reviewers with steps they can take to test that the PR does what it says it does.

1. Install new csv downloader dep `npm i`
2. Build `npm run dev`
3. Login as admin
4. Navigate to `/admin/pool-candidates`
5. Select some rows
6. Note, not query is ran
7. Attempt both download/print
8. Confirm query is ran and expected output document is created
9. Navigate to `/admin/users`
10. Repeat steps 5-8
11. Navigate to an application (`/admin/candidates/{poolCandidateId}/application`)
12. Click "Print application"
13. Confirm it still functions as expected
14. Navigate to a user profile (`/admin/users/{userId}/profile`)
15. Click "Print profile"
16. Select either option
17. Confirm it still functions as expected

### CODE REVIEW

**Comment by tristan-orourke** [comment]
Hey, based on some further dialog with recruitment team, do you mind if I add another Acceptance Criteria?

Basically, the request for data sometimes fails with a server error if you request too many candidates at once. The graphql response is simply `{ "message": "Server Error" }`. Right now this seems to fail silently. Could you make sure we throw a toast if the request fails like this, ideally with a suggestion to reduce the number of items selected?

**Comment by esizer** [comment]
> Basically, the request for data sometimes fails with a server error if you request too many candidates at once. The graphql response is simply `{ "message": "Server Error" }`. Right now this seems to fail silently. Could you make sure we throw a toast if the request fails like this, ideally with a suggestion to reduce the number of items selected?

Yup! I added it [add error message for too many rows on download](https://github.com/GCTC-NTGC/gc-digital-talent/pull/8970/commits/9b8e15cc070077c5d79a88321c4648c3062ade00). I'm not happy with the actual message but I was just following the same pattern for the "no rows" error. Any suggestions on a better message? 🤔

**Comment by tristan-orourke** [comment]
I have to admit, it's not obvious to me that this icon is meant to act like a loading spinner. Without moving, it looks more like a "retry" button. 😕 

image

**Review by petertgiles — CHANGES_REQUESTED** [review]
Nice!  This is a pretty major refactor, actually.  Turned out great.  :1st_place_medal:  I noticed one bug in the test but overall works really well.

I found it a bit weird that the UI throws up spinners beside both links regardless of which one was clicked.  I know there was no design provided so I won't block on that but I think it would make a bit more sense if there was a single shared spinner associated with the entire footer, maybe.

File: apps/e2e/cypress/e2e/admin/admin-workflows.cy.ts [hunk 1]
```diff
@@ -104,7 +104,7 @@ describe("Admin Workflow Tests", () => {
       .and("be.visible");
   });
 
-  it("Selects a user and downloads a CSV", () => {
+  it.only("Selects a user and downloads a CSV", () => {
```
[the remark this answers is not in the document]
Comment (esizer) [reply apps/e2e/cypress/e2e/admin/admin-workflows.cy.ts]: Oof, goo catch! Fixed in [undo single admin workflow test debug](https://github.com/GCTC-NTGC/gc-digital-talent/pull/8970/commits/9e7bb34ff34c6fedf2789e4ffedb2078d4d3873d)

**Comment by esizer** [comment]
> I found it a bit weird that the UI throws up spinners beside both links regardless of which one was clicked. I know there was no design provided so I won't block on that but I think it would make a bit more sense if there was a single shared spinner associated with the entire footer, maybe.

I agree, it was a simple fix so I did it in [only show loading icon for specific action](https://github.com/GCTC-NTGC/gc-digital-talent/pull/8970/commits/5a36bd440d5ba00967b2a11ad9b018504baae1db)

**Review by petertgiles — APPROVED** [review]
Looks great!

File: apps/web/src/components/PoolCandidatesTable/PoolCandidatesTable.tsx [hunk 2]
```diff
@@ -335,41 +344,30 @@ const PoolCandidatesTable = ({
...
+      })
+      .toPromise()
+      .then((result) => {
+        const poolCandidates: PoolCandidate[] = unpackMaybes(
+          result.data?.poolCandidates,
+        );
 
-  const selectedCandidates =
-    selectedCandidatesData?.poolCandidates.filter(notEmpty) ?? [];
+        if (result.error) {
```
[the remark this answers is not in the document]
Comment (esizer) [reply apps/web/src/components/PoolCandidatesTable/PoolCandidatesTable.tsx:361]: oh my, yeah sorry. I totally read the first comment incorrectly 🤦‍♀️ I thought it said "no candidates" not "too many" 😬

[the remark this answers is not in the document]
Comment (esizer) [reply apps/web/src/components/PoolCandidatesTable/PoolCandidatesTable.tsx:362]: I am having trouble reproducing this exact type of error. May need some help with that for testing but, I added a catch statement and used the generic unknown error message (which _should_ cover it).

**Review by tristan-orourke — APPROVED** [review]
Thanks! 👍

**Pull request merged**